src/pages/index.js

```diff
@@ -29,50 +29,165 @@ import {
+  }
+);
+// =============================================================================
+// Handle Clicks
+// =============================================================================
+
+function handleLikeClick() {
+  if (this._likes) {
+    api
+      .unlikeCard(this._id)
+      .then((res) => {
+        this.setLikes(res.isLiked);
+      })
+      .catch((error) => {
+        console.log(`An error has occured ${error}`);
+      });
+  } else {
+    api
+      .likeCard(this._id)
+      .then((res) => {
+        this.setLikes(res.isLiked);
+      })
+      .catch((err) => console.error(err));
+  }
+}
+
+function handleCardClick(data) {
+  previewImagePopup.open(data);
+}
+
+const deleteCardPopup = new PopupWithConfirmation("#delete-card-modal");
+deleteCardPopup.setEventListeners();
+
+function handleDeleteClick(card) {
+  deleteCardPopup.open();
+  deleteCardPopup.setSubmitAction(() => {
+    deleteCardPopup.setLoading(true);
+    api
+      .deleteCard(card.getId())
+      .then(() => {
+        card.remove();
+        deleteCardPopup.close();
+      })
+      .catch((err) => console.error(err))
+      .finally(() => {
+        deleteCardPopup.setLoading(false, "Yes");
+      });
+  });
+}
+
+function createCard({ name, link, isLiked, _id, userId, ownerId }) {
   const cardElement = new Card(
-    { name, link },
+    { name, link, isLiked, _id, userId, ownerId },
+    userID,
     "#card-template",
-    ({ name, link }) => {
-      previewImagePopup.open({ name, link });
-    }
+    handleCardClick,
+    handleDeleteClick,
+    handleLikeClick
   );
   return cardElement.getView();
 }
-
+// =============================================================================
 // Preview Popup
+// =============================================================================
 
 const previewImagePopup = new PopupWithImage({
   popupSelector: "#preview-modal",
+  handleFormSubmit: handleProfileEditSubmit,
 });
 previewImagePopup.setEventListeners();
 
-//Profile Popup
-
-const newUser = new UserInfo(".profile__title", ".profile__description");
+// =============================================================================
+// Profile Popup
+// =============================================================================
 
-function handleProfileEditSubmit({ title, description }) {
-  newUser.setUserInfo(title, description);
+function handleProfileEditSubmit({ name, about }) {
+  newUser.setUserInfo(name, about);
   profilePopup.close();
 }
 
-const profilePopup = new PopupWithForm(
-  "#profile-edit-modal",
-  handleProfileEditSubmit
-);
+const profilePopup = new PopupWithForm({
+  popupSelector: "#profile-edit-modal",
+  handleFormSubmit: (inputValues) => {
+    profilePopup.showLoading();
+    api
+      .editProfile(inputValues)
+      .then(() => {
+        userInfo.setUserInfo(inputValues);
+        profilePopup.close();
+      })
+      .catch((err) => console.error(err))
+      .finally(() => {
+        profilePopup.hideLoading();
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```